### PR TITLE
Secure visualizer app with authentification

### LINKED ISSUE #1073: Secure visualizer app with authentication and https
## What and why
Anyone with the URL who is logged in to the MGH network can access it.

Aaron wants this app to either be:
1. secured so it requires a username and password login and signed with SSL certificate
2. run local-only and not served over the MGH network

I think option 2 is simpler and faster.

Also, I think it is a good idea to remove the functionality to download selected data. Our pipeline doesn't support that exported data format. It is better to keep the functionality of the `visualizer` focused, and not introduce additional data formats.

### PR DESCRIPTION
**Pull request by epalto**
## Issues closed
Closes #1073 

## Changelog
- [x] Shell prints a token to authenticate users when using the visualizer

## Checklist
- [x] The following are complete:
    - Added or revised tests if appropriate.
    - If new packages are installed, `requirements.txt` is updated.
    - If documentation should be updated, a new issue was made in `ml4c3-wiki`.
    - The PR name describes its content and does not contain "Ref #".
    - The PR does not add PHI.

### CODE REVIEW

**Comment by erikr** [comment]
Aaron said the web app should only be run locally.

Eric,
1. Does this change interrupt your or anyone else's workflow?
2. Is the change be difficult to make? I think it would involve removing some code and actually simplify the recipe.

Also, did you find how to push the updated Docker images to ghcr.io?

**Comment by erikr** [comment]
Tests fail on `master`. Will investigate.

**Comment by epalto** [comment]
Thanks Erik! This PR has the changes needed to secure the visualizer locally so no further work should be done for Aaron's request. I also built and pushed the docker image but it seems that whenever the image is rebuilt, tests fail

**Review by erikr — CHANGES_REQUESTED** [review]
I tested the functionality on the new branch. Nice work!

Although the app is served to `0.0.0.0`, can we print the address `http://$MACHINE_NAME.partners.org:8888` so the user can more easily get to the app?

Can you remind me why `-m /media/mad3/hd5` is required when `./scripts/run.sh` mounts the entirety of `/media` by default?

**Comment by StevenSong** [comment]
When Aaron says that we must run locally, what does that mean? that we shouldnt be able to access it from e.g. everest.partners.org? Should we have it setup such that we must ssh tunnel to the machine to access it remotely or remote desktop to the machine to access it?

**Comment by erikr** [comment]
The most secure solution is to not be able to access the app at `everest.partners.org` from any machine on the MGH network.

I think SSH tunneling could work. But I dislike remote desktop.

Thoughts?

**Comment by StevenSong** [comment]
i see, so it must not be served to everest.partners.org - this is a more complex ask. the solutions i can think of are:
- change the firewall on everest to restrict the port to only the local machine (not even local network) - the issue with this is it is difficult to generalize through code/automation i.e. if we want to run the visualizer on olympus, we ought to modify the firewall on olympus. we ought to also restrict the port for the visualizer then
- the other solution may just be to have the webapp itself block connections from external IP addresses, which should be doable from the visualizer code - I'm not sure if this actually works with ssh tunneling but something to test

**Comment by StevenSong** [comment]
https://stackoverflow.com/questions/22251038/how-to-limit-access-to-flask-for-a-single-ip-address

this SO actually has solutions to both kinds of questions

actually another possibility is to configure the firewall within our docker container via the run script to restrict access - this might be the best option (OS level restriction and portable because we use docker) but still more complex than just doing it in Flask

**Comment by epalto** [comment]
Should those restrictions be applied to Jupyter Notebooks too? If so, it may be better to configurate it on docker so that it secures both things at the same time

>Can you remind me why -m /media/mad3/hd5 is required when ./scripts/run.sh mounts the entirety of /media by default?

When you launch the visualizer, this path is automatically written on the hd5_path input so that the user doesn't have to type it every time.

File: visualizer/run.py [hunk 1]
```diff
@@ -11,6 +15,12 @@
 def run(debug, port, address, files_dir, options):
     app.title = "HD5 visualizer"
 
+    random_tocken = str(uuid4())
```
Comment (erikr) [inline visualizer/run.py:18]:
```suggestion
    random_token = str(uuid4())
```

File: visualizer/run.py [hunk 2]
```diff
@@ -11,6 +15,12 @@
...
+    dash_auth.BasicAuth(
+        app,
+        {"aguirrelab": random_tocken},
```
Comment (erikr) [inline visualizer/run.py:21]:
```suggestion
        {"aguirrelab": random_token},
```

File: visualizer/run.py [hunk 3]
```diff
@@ -28,12 +38,21 @@ def run(debug, port, address, files_dir, options):
...
 
+    print(
+        f"""
+    ****************
+    To log in use:
 
-def run_server(args):
-    run(args.debug, args.port, args.address, args.tensors, args.options_file)
+    \t Username: aguirrelab
+    \t Password: {random_tocken}
```
Comment (erikr) [inline visualizer/run.py:48]:
```suggestion
    \t Password: {random_token}
```

**Comment by erikr** [comment]
Why the import statements inside of `run`? Why not put them at the top of the script?

**Comment by epalto** [comment]
> What do you think of the following syntax?

The reason why we chose this notation is for synchronization with Docker. 
The `--port` argument is not free because the chosen port needs to be linked with the container port. By default, the container port links the port 8888 so if you run:
```
./scripts/run.sh -v \
...
--port 8010 \
...
```
the visualizer will run inside Docker on port 8010, which won't be accessible from outside the container.

The same thing happens with `--tensors`: the path is input with the `-m` option to make sure that whichever path is chosen, it gets mounted to the container. For instance, if you had data in `/data/hd5`,  the syntax ` ./scripts/run.sh -m /data/hd5 -v` would both mount that path and launch visualizer in it, while the syntax `--tensors /data/hd5` would not mount the path so the visualizer wouldn't find the files.

Anyway, this may need to be changed when once we have different paths for BM and EDW.

> Why the import statements inside of run? 

That's because some code needs to be run before importing those files. In particular, the layout can be customized with the `--options_file` so the configuration needs to be loaded before creating the layout and setting the callbacks.

**Comment by erikr** [comment]
> The reason why we chose this notation is for synchronization with Docker.

I understand, thanks for explaining. Makes sense.

> The same thing happens with --tensors: the path is input with the -m option to make sure that whichever path is chosen, it gets mounted to the container. For instance, if you had data in /data/hd5, the syntax ./scripts/run.sh -m /data/hd5 -v would mount that path and launch visualizer on it, while the syntax --tensors /data/hd5 would not mount the path so the visualizer wouldn't find the files.

I believe `-m` in `run.sh` recursively mounts at that path. So when we mount via `-m /data`, Docker has access to `/data/hd5`, `/data/edw`, etc.. and `--tensors /data/any_subdirectory` works.

When we call `train` recipes, we don't mount specific subfolders. We by default mount `-m /storage/shared`, and can access and train on data in `/storage/shared/ecg`.

@StevenSong am I missing something here?

**Comment by epalto** [comment]
That's right, the `-m /media/hd5/` is redundant because the `/media/` directory is mounted by default and it includes all subdirectories, but the visualizer needs the specific path where it looks for the file. It is only redundant if your data is inside the `/media` directory because this one is mounted by default, but if you want to add data from another directory, say the `/data/hd5` directory, it won't be redundant

**Comment by erikr** [comment]
> ...but it's a way to specify the exact path where the .hd5 files are stored, which is needed by the visualizer

I understand, thanks for clarifying.

I propose a future improvement (out of scope for this PR though) would be to not require the user to mount the exact path in the run script via `-m`. The app would have a dropdown menu that lists every directory found at the `-m` path (identified via `os.walk`), and the user can select the parent dir containing HD5s they wish to load and visualize data from.

**Comment by epalto** [comment]
Sounds good! How about we incorporate that when we add the modifications for an EDW path and a BM path (Separate EDW and Bedmaster milestone)?

**Comment by erikr** [comment]
> Sounds good! How about we incorporate that when we add the modifications for an EDW path and a BM path (issue #1126)?

Great, thank you for opening that new issue. I appreciate it!

**Review by erikr — APPROVED** [review]

**Comment by erikr** [comment]
Discussed current authentication method. We believe this is sufficiently secure. Can revisit and devise other authentication protocols in the future if necessary.

**Pull request merged**